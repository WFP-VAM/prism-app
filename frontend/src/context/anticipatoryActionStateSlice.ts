import { PayloadAction, createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import Papa from 'papaparse';
import { LayerDefinitions } from 'config/utils';
import { AnticipatoryActionLayerProps } from 'config/types';
import type { CreateAsyncThunkTypes, RootState } from './store';

export const allWindowsKey = 'All';

// na/ny are not actually found in CSV, but defined not to cause confusion when calling the functions
export const AAcategory = ['na', 'ny', 'Leve', 'Moderado', 'Severo'] as const;
export type AACategoryType = typeof AAcategory[number];

export const AAPhase = ['na', 'ny', 'Ready', 'Set'] as const;
export type AAPhaseType = typeof AAPhase[number];

export interface AnticipatoryActionData {
  category: AACategoryType;
  district: string;
  index: string;
  month: string;
  phase: AAPhaseType;
  probability: string;
  trigger: string;
  triggerNB: string;
  triggerType: string;
  type: string;
  windows: string;
  yearOfIssue: string;
  date: string;
}

export const AAlayerKey = 'anticipatory_action';

const AACSVKeys: [string, string][] = [
  ['Category', 'category'],
  ['District', 'district'],
  ['Index', 'index'],
  ['Month', 'month'],
  ['Phase', 'phase'],
  ['Probability', 'probability'],
  ['Trigger', 'trigger'],
  ['Trigger_nb', 'triggerNB'],
  ['Trigger_type', 'triggerType'],
  ['Type', 'type'],
  ['Windows', 'windows'],
  ['Year_of_issue', 'yearOfIssue'],
];

function transform(data: any[], keys: [string, string][]) {
  const parsed = data
    .filter(x => !!x[keys[0][0]]) // filter empty rows
    .map(obj => {
      const entries = keys.map(k => [k[1], obj[k[0]]]);
      const month = obj.Month.padStart(2, '0');
      const year = obj.Year_of_issue.split('-')[0];

      // TODO: update this once CSV date format is revised
      const actualYear = Number(month) < 5 ? String(Number(year) + 1) : year;
      const date = `${actualYear}-${month}-01`;
      return Object.fromEntries([...entries, ['date', date]]);
    }) as AnticipatoryActionData[];

  const windows = [...new Set(parsed.map(x => x.windows))];
  // eslint-disable-next-line fp/no-mutating-methods
  const availableDates = [
    ...new Set(parsed.map(x => new Date(x.date).getTime())),
  ].sort();

  const groupedByDistrict = new Map<string, AnticipatoryActionData[]>();

  parsed.forEach(x => {
    const { district } = x;
    const rows = groupedByDistrict.get(district);
    groupedByDistrict.set(district, rows ? [...rows, x] : [x]);
  });

  const sortFn = (a: AnticipatoryActionData, b: AnticipatoryActionData) => {
    const aDate = new Date(a.date).valueOf();
    const bDate = new Date(b.date).valueOf();
    if (aDate > bDate) {
      return -1;
    }
    if (bDate > aDate) {
      return 1;
    }
    const aCatIndex = AAcategory.findIndex(x => x === a.category);
    const bCatIndex = AAcategory.findIndex(x => x === b.category);
    if (aCatIndex > bCatIndex) {
      return -1;
    }
    if (bCatIndex > aCatIndex) {
      return 1;
    }
    return 0;
  };

  const result = Object.fromEntries(
    // eslint-disable-next-line fp/no-mutating-methods
    Array.from(groupedByDistrict.entries()).map(x => [x[0], x[1].sort(sortFn)]),
  );

  const monitoredDistricts = Object.keys(result);

  return { data: result, windows, availableDates, monitoredDistricts };
}

type AnticipatoryActionState = {
  data: {
    [k: string]: AnticipatoryActionData[];
  };
  availableDates?: number[];
  monitoredDistricts: string[];
  windows: string[];
  selectedWindow: string;
  categoryFilters: Record<AACategoryType, boolean>;
  loading: boolean;
  error: string | null;
};

const initialState: AnticipatoryActionState = {
  data: {},
  windows: [],
  availableDates: undefined,
  monitoredDistricts: [],
  selectedWindow: allWindowsKey,
  categoryFilters: {
    Severo: true,
    Moderado: true,
    Leve: true,
    na: false,
    ny: false,
  },
  loading: false,
  error: null,
};

export const loadAAData = createAsyncThunk<
  {
    data: {
      [k: string]: AnticipatoryActionData[];
    };
    windows: string[];
    availableDates: number[];
    monitoredDistricts: string[];
  },
  undefined,
  CreateAsyncThunkTypes
>('anticipatoryActionState/loadAAData', async () => {
  const layer = LayerDefinitions[AAlayerKey] as AnticipatoryActionLayerProps;

  const url = layer.baseUrl;
  return new Promise<any>((resolve, reject) =>
    Papa.parse(url, {
      header: true,
      download: true,
      complete: results => resolve(transform(results.data, AACSVKeys)),
      error: error => reject(error),
    }),
  );
});

export const anticipatoryActionStateSlice = createSlice({
  name: 'anticipatoryActionState',
  initialState,
  reducers: {
    setSelectedWindow: (state, { payload }: PayloadAction<string>) => ({
      ...state,
      selectedWindow: payload,
    }),
    setCategoryFilters: (
      state,
      {
        payload,
      }: PayloadAction<
        Partial<Record<'Leve' | 'Moderado' | 'Severo', boolean>>
      >,
    ) => ({
      ...state,
      categoryFilters: { ...state.categoryFilters, ...payload },
    }),
  },
  extraReducers: builder => {
    builder.addCase(loadAAData.fulfilled, (state, { payload }) => ({
      ...state,
      loading: false,
      data: payload.data,
      windows: payload.windows,
      availableDates: payload.availableDates,
      monitoredDistricts: payload.monitoredDistricts,
    }));

    builder.addCase(loadAAData.rejected, (state, action) => ({
      ...state,
      loading: false,
      error: action.error.message
        ? action.error.message
        : action.error.toString(),
    }));

    builder.addCase(loadAAData.pending, ({ error, ...state }) => ({
      ...state,
      error: null,
      loading: true,
    }));
  },
});

// export selectors
export const AnticipatoryActionDataSelector = (state: RootState) =>
  state.anticipatoryActionState.data;

export const AnticipatoryActionWindowsSelector = (state: RootState) =>
  state.anticipatoryActionState.windows;

export const AnticipatoryActionAvailableDatesSelector = (state: RootState) =>
  state.anticipatoryActionState.availableDates;

export const AASelectedWindowSelector = (state: RootState) =>
  state.anticipatoryActionState.selectedWindow;

export const AACategoryFiltersSelector = (state: RootState) =>
  state.anticipatoryActionState.categoryFilters;

// export actions
export const {
  setSelectedWindow,
  setCategoryFilters,
} = anticipatoryActionStateSlice.actions;

export default anticipatoryActionStateSlice.reducer;
