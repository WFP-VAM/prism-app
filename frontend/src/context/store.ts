import {
  combineReducers,
  configureStore,
  getDefaultMiddleware,
} from '@reduxjs/toolkit';
import mapStateReduce from './mapStateSlice';
import serverStateReduce from './serverStateSlice';
import tableStateReduce from './tableStateSlice';
import tooltipStateReduce from './tooltipStateSlice';
import notificationStateReduce, {
  errorToNotificationMiddleware,
} from './notificationStateSlice';
import analysisResultStateReduce from './analysisResultStateSlice';
import mapSelectionLayerStateReduce from './mapSelectionLayerStateSlice';
import mapBoundaryInfoStateReduce from './mapBoundaryInfoStateSlice';
import datasetResultStateReduce from './datasetStateSlice';
import mapTileLoadingStateReduce from './mapTileLoadingStateSlice';
import leftPanelStateReduce from './leftPanelStateSlice';
import opacityStateReduce from './opacityStateSlice';
import anticipatoryActionDroughtStateReduce from './anticipatoryAction/AADroughtStateSlice';
import anticipatoryActionStormStateReduce from './anticipatoryAction/AAStormStateSlice';

const reducer = combineReducers({
  mapState: mapStateReduce,
  serverState: serverStateReduce,
  tableState: tableStateReduce,
  tooltipState: tooltipStateReduce,
  analysisResultState: analysisResultStateReduce,
  notificationState: notificationStateReduce,
  mapSelectionLayerStateSlice: mapSelectionLayerStateReduce,
  mapBoundaryInfoState: mapBoundaryInfoStateReduce,
  datasetState: datasetResultStateReduce,
  mapTileLoadingState: mapTileLoadingStateReduce,
  leftPanelState: leftPanelStateReduce,
  opacityState: opacityStateReduce,
  anticipatoryActionDroughtState: anticipatoryActionDroughtStateReduce,
  anticipatoryActionStormState: anticipatoryActionStormStateReduce,
});

export const store = configureStore({
  reducer,
  middleware: getDefaultMiddleware({
    // TODO: Instead of snoozing this check, we might want to
    // serialize the state
    serializableCheck: false,
    immutableCheck: {
      // do not check the following state branches for accidental
      // mutations. This saves a lot of time in dev mode (but has no
      // impact on production builds).
      ignoredPaths: [
        'mapState.boundaryRelations',
        'mapState.layersData',
        'analysisResultState.result',
        'serverState.availableDates',
      ],
    },
  }).concat(errorToNotificationMiddleware),
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof reducer>;

export type ThunkApi = {
  dispatch: AppDispatch;
  getState: () => RootState;
  // can add 'extra' and 'rejectValue' here if we ever need them
  // https://redux-toolkit.js.org/usage/usage-with-typescript#createasyncthunk
};

// Used as the third type definition for typing createAsyncThunk
// e.g. createAsyncThunk<ReturnType, HandlerType, CreateAsyncThunkTypes>
export type CreateAsyncThunkTypes = {
  dispatch: AppDispatch;
  state: RootState;
};
