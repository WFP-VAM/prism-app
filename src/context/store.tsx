import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import filtersReduce from './filters/filtersSlice';

export const store = configureStore({
  reducer: {
    filters: filtersReduce,
  },
  // TODO: Instead of snoozing this check, we might want to
  // serialize the state
  middleware: getDefaultMiddleware({
    serializableCheck: false,
  }),
});

export type RootState = ReturnType<typeof store.getState>;
