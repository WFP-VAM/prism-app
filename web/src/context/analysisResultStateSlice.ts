import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
import { Position } from 'geojson';
import { get } from 'lodash';
import type { CreateAsyncThunkTypes, RootState } from './store';
import {
  AggregationOperations,
  AsyncReturnType,
  BoundaryLayerProps,
  NSOLayerProps,
  ThresholdDefinition,
  WMSLayerProps,
} from '../config/types';
import {
  AnalysisResult,
  ApiData,
  BaselineLayerData,
  checkBaselineDataLayer,
  fetchApiData,
  generateFeaturesFromApiData,
  scaleAndFilterAggregateData,
} from '../utils/analysis-utils';
import { getWCSLayerUrl } from './layers/wms';
import { getBoundaryLayerSingleton } from '../config/utils';
import { Extent } from '../components/MapView/Layers/raster-utils';
import { layerDataSelector } from './mapStateSlice/selectors';
import { LayerData, LayerDataParams, loadLayerData } from './layers/layer-data';
import { DataRecord, NSOLayerData } from './layers/nso';
import { BoundaryLayerData } from './layers/boundary';

const ANALYSIS_API_URL = 'https://prism-api.ovio.org/stats'; // TODO both needs to be stored somewhere
const DEV_ADMIN_BOUNDARIES_URL =
  'https://prism-admin-boundaries.s3.us-east-2.amazonaws.com/mng_admin_boundaries.json';

type AnalysisResultState = {
  result?: AnalysisResult;
  error?: string;
  isLoading: boolean;
  isMapLayerActive: boolean;
};

export type TableRow = {
  localName: string;
  name: string;
  baselineValue: DataRecord['value'];
  coordinates?: Position;
} & { [k in AggregationOperations]?: number };

const initialState: AnalysisResultState = {
  isLoading: false,
  isMapLayerActive: true,
};

/* Gets a public URL for the admin boundaries used by this application.
 *
 * If the application is in development, localhost is not accessible publicly.
 * Therefore, we will return a pre-set constant to be used for development.
 *
 * If the application is in production, we will attempt to construct a public URL that the backend should be able to access.
 */
function getAdminBoundariesURL() {
  const isProduction = process.env.NODE_ENV === 'production';
  const adminBoundariesPath = getBoundaryLayerSingleton().path;
  if (!isProduction) {
    return DEV_ADMIN_BOUNDARIES_URL;
  }
  // already a remote location, so return it.
  if (adminBoundariesPath.startsWith('http')) {
    return adminBoundariesPath;
  }
  // the regex here removes the dot at the beginning of a path, if there is one.
  // e.g the path might be ' ./data/xxx '  instead of ' /data/xxx '
  return window.location.origin + adminBoundariesPath.replace(/^\./, '');
}

function generateTableFromApiData(
  statistic: AggregationOperations,
  aggregateData: AsyncReturnType<typeof fetchApiData>, // data from api
  // admin layer, both props and data. Aimed to closely match LayerData<BoundaryLayerProps>
  {
    layer: adminLayer,
    data: adminLayerData,
  }: { layer: BoundaryLayerProps; data: BoundaryLayerData },
  // baseline layer, both props and data
  {
    layer: { adminLevel },
    data: { layerData: baselineLayerData },
  }: { layer: NSOLayerProps; data: NSOLayerData },
): TableRow[] {
  // find the key that will let us reference the names of the bounding boxes. We get the one corresponding to the specific level of baseline, or the first if we fail.
  const adminLevelName =
    adminLayer.adminLevelNames[adminLevel - 1] || adminLayer.adminLevelNames[0];
  // for local name too.
  const adminLevelLocalName =
    adminLayer.adminLevelLocalNames[adminLevel - 1] ||
    adminLayer.adminLevelLocalNames[0];

  return aggregateData.map(row => {
    // find feature (a cell on the map) from admin boundaries json that closely matches this api row.
    // we decide it matches if the feature json has the same name as the name for this row.
    // once we find it we can get the corresponding local name.
    const featureBoundary = adminLayerData.features.find(
      feature => feature.properties?.[adminLevelName] === row[adminLevelName],
    );

    const name: string =
      featureBoundary?.properties?.[adminLevelName] || 'No Name';
    const localName: string =
      featureBoundary?.properties?.[adminLevelLocalName] || 'No Name';

    // we are searching the data of baseline layer to find the data associated with this feature
    // adminKey here refers to a specific feature (could be several) where the data is attached to.
    const baselineValue =
      baselineLayerData.find(({ adminKey }) => {
        // we only check startsWith because the adminCode grows longer the deeper the level.
        // For example, 34 is state and 14 is district, therefore 3414 is a specific district in a specific state.
        // if this baseline layer only focuses on a higher level (just states) it would only contain 34, but every feature is very specific (uses the full number 3414)
        // therefore checking the start will cover all levels.
        return featureBoundary?.properties?.[adminLayer.adminCode].startsWith(
          adminKey,
        );
      })?.value || 'No Data';

    const tableRow: TableRow = {
      name,
      localName,
      [statistic]: get(row, statistic, 0),
      baselineValue,
      coordinates: (featureBoundary?.geometry as any).coordinates[0][0][0], // TODO likely will not keep
    };
    return tableRow;
  });
}

export type AnalysisDispatchParams = {
  baselineLayer: NSOLayerProps;
  hazardLayer: WMSLayerProps;
  extent: Extent;
  threshold: ThresholdDefinition;
  date: ReturnType<Date['getTime']>; // just a hint to developers that we give a date number here, not just any number
  statistic: AggregationOperations; // we might have to deviate from this if analysis accepts more than what this enum provides
};

export const requestAndStoreAnalysis = createAsyncThunk<
  AnalysisResult,
  AnalysisDispatchParams,
  CreateAsyncThunkTypes
>('analysisResultState/requestAndStoreAnalysis', async (params, api) => {
  const {
    hazardLayer,
    date,
    baselineLayer,
    extent,
    statistic,
    threshold,
  } = params;
  const baselineData = layerDataSelector(baselineLayer.id)(
    api.getState(),
  ) as LayerData<NSOLayerProps>;
  const adminBoundaries = getBoundaryLayerSingleton();
  const adminBoundariesData = layerDataSelector(adminBoundaries.id)(
    api.getState(),
  ) as LayerData<BoundaryLayerProps>;

  if (!adminBoundariesData) {
    throw new Error('Boundary Layer not loaded!');
  }
  // we force group_by to be defined with &
  // eslint-disable-next-line camelcase
  const apiRequest: ApiData & { group_by: string } = {
    geotiff_url: getWCSLayerUrl({
      layer: hazardLayer,
      date,
      extent,
    }),
    zones_url: getAdminBoundariesURL(),
    group_by:
      adminBoundaries.adminLevelNames[baselineLayer.adminLevel - 1] ||
      adminBoundaries.adminLevelNames[0],
  };
  const aggregateData = scaleAndFilterAggregateData(
    await fetchApiData(ANALYSIS_API_URL, apiRequest),
    hazardLayer,
    statistic,
    threshold,
  );
  let loadedAndCheckedBaselineData: BaselineLayerData;
  // if the baselineData doesn't exist, lets load it, otherwise check then load existing data.
  // similar code can be found at impact.ts
  if (!baselineData) {
    const {
      payload: { data },
    } = (await api.dispatch(
      loadLayerData({ layer: baselineLayer, extent } as LayerDataParams<
        NSOLayerProps
      >),
    )) as { payload: { data: unknown } };

    // eslint-disable-next-line fp/no-mutation
    loadedAndCheckedBaselineData = checkBaselineDataLayer(
      baselineLayer.id,
      data,
    );
  } else {
    // eslint-disable-next-line fp/no-mutation
    loadedAndCheckedBaselineData = checkBaselineDataLayer(
      baselineLayer.id,
      baselineData.data,
    );
  }

  const features = generateFeaturesFromApiData(
    aggregateData,
    loadedAndCheckedBaselineData,
    apiRequest.group_by,
    statistic,
  );

  const tableRows: TableRow[] = generateTableFromApiData(
    statistic,
    aggregateData,
    adminBoundariesData,
    { layer: baselineLayer, data: loadedAndCheckedBaselineData },
  );

  return new AnalysisResult(
    tableRows,
    {
      ...adminBoundariesData.data,
      features,
    },
    hazardLayer,
    baselineLayer,
    statistic,
    threshold,
    // we never use the raw api data besides for debugging. So lets not bother saving it in Redux for production
    process.env.NODE_ENV === 'production' ? undefined : aggregateData,
  );
});

export const analysisResultSlice = createSlice({
  name: 'analysisResultState',
  initialState,
  reducers: {
    setIsMapLayerActive: (state, { payload }: PayloadAction<boolean>) => ({
      ...state,
      isMapLayerActive: payload,
    }),
    clearAnalysisResult: state => ({
      ...state,
      result: undefined,
    }),
  },
  extraReducers: builder => {
    builder.addCase(
      requestAndStoreAnalysis.fulfilled,
      (
        { result, ...rest },
        { payload }: PayloadAction<AnalysisResult>,
      ): AnalysisResultState => ({
        ...rest,
        isLoading: false,
        result: payload,
      }),
    );

    builder.addCase(
      requestAndStoreAnalysis.rejected,
      (state, action): AnalysisResultState => ({
        ...state,
        isLoading: false,
        error: action.error.message
          ? action.error.message
          : action.error.toString(),
      }),
    );

    builder.addCase(
      requestAndStoreAnalysis.pending,
      (state): AnalysisResultState => ({
        ...state,
        isLoading: true,
      }),
    );
  },
});

// Getters
export const analysisResultSelector = (
  state: RootState,
): AnalysisResult | undefined => state.analysisResultState.result;

export const isAnalysisLoadingSelector = (state: RootState): boolean =>
  state.analysisResultState.isLoading;

export const isAnalysisLayerActiveSelector = (state: RootState): boolean =>
  state.analysisResultState.isMapLayerActive;

// Setters
export const {
  setIsMapLayerActive,
  clearAnalysisResult,
} = analysisResultSlice.actions;

export default analysisResultSlice.reducer;
