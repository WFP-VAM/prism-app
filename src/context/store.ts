import {
  combineReducers,
  configureStore,
  getDefaultMiddleware,
} from '@reduxjs/toolkit';

import mapStateReduce from './mapStateSlice';
import serverStateReduce from './serverStateSlice';
import tableStateReduce from './tableStateSlice';
import tooltipStateReduce from './tooltipStateSlice';
import notificationStateReduce, {
  errorToNotificationMiddleware,
} from './notificationStateSlice';
import analysisResultStateSlice from './analysisResultStateSlice';

const reducer = combineReducers({
  mapState: mapStateReduce,
  serverState: serverStateReduce,
  tableState: tableStateReduce,
  tooltipState: tooltipStateReduce,
  analysisResultState: analysisResultStateSlice,
  notificationState: notificationStateReduce,
});

export const store = configureStore({
  reducer,
  // TODO: Instead of snoozing this check, we might want to
  // serialize the state
  middleware: getDefaultMiddleware({
    serializableCheck: false,
    immutableCheck: {
      ignoredPaths: ['mapState.layersData', 'analysisResultState.result'],
    },
  }).concat(errorToNotificationMiddleware),
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof reducer>;

export type ThunkApi = {
  dispatch: AppDispatch;
  getState: () => RootState;
  // can add 'extra' and 'rejectValue' here if we ever need them
  // https://redux-toolkit.js.org/usage/usage-with-typescript#createasyncthunk
};

// Used as the third type definition for typing createAsyncThunk
// e.g. createAsyncThunk<ReturnType, HandlerType, CreateAsyncThunkTypes>
export type CreateAsyncThunkTypes = {
  dispatch: AppDispatch;
  state: RootState;
};
