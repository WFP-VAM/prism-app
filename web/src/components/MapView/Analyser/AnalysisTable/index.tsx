import React, { useState } from 'react';
import {
  createStyles,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TablePagination,
  TableRow,
  withStyles,
  WithStyles,
} from '@material-ui/core';
import { useDispatch } from 'react-redux';
import { TableRow as AnalysisTableRow } from '../../../../context/analysisResultStateSlice';
import { showPopup } from '../../../../context/tooltipStateSlice';
import { Column } from '../../../../utils/analysis-utils';

function AnalysisTable({ classes, tableData, columns }: AnalysisTableProps) {
  const [page, setPage] = useState(0);
  const [rowsPerPage, setRowsPerPage] = useState(10);

  const dispatch = useDispatch();

  const handleChangePage = (event: unknown, newPage: number) => {
    setPage(newPage);
  };

  const handleChangeRowsPerPage = (
    event: React.ChangeEvent<HTMLInputElement>,
  ) => {
    setRowsPerPage(+event.target.value);
    setPage(0);
  };

  return (
    <div>
      <TableContainer className={classes.tableContainer}>
        <Table stickyHeader aria-label="sticky table">
          <TableHead>
            <TableRow>
              {columns.map(column => (
                <TableCell key={column.id} className={classes.tableHead}>
                  {column.label}
                </TableCell>
              ))}
            </TableRow>
          </TableHead>
          <TableBody>
            {tableData
              .slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage)
              .map(row => {
                return (
                  <TableRow
                    hover
                    role="checkbox"
                    tabIndex={-1}
                    key={row.name}
                    onClick={() => {
                      // TODO if we decide to keep, add popup data?
                      if (row.coordinates) {
                        dispatch(
                          showPopup({
                            coordinates: row.coordinates,
                            locationName: row.name,
                          }),
                        );
                      }
                    }}
                    style={{ cursor: row.coordinates ? 'pointer' : 'none' }}
                  >
                    {columns.map(column => {
                      const value = row[column.id];
                      return (
                        <TableCell key={column.id}>
                          {column.format && typeof value === 'number'
                            ? column.format(value)
                            : value}
                        </TableCell>
                      );
                    })}
                  </TableRow>
                );
              })}
          </TableBody>
        </Table>
      </TableContainer>
      <TablePagination
        rowsPerPageOptions={[10, 25, 100]}
        component="div"
        count={tableData.length}
        rowsPerPage={rowsPerPage}
        page={page}
        onChangePage={handleChangePage}
        onChangeRowsPerPage={handleChangeRowsPerPage}
      />
    </div>
  );
}

const styles = () =>
  createStyles({
    tableContainer: {
      border: '2px solid',
      width: '700px',
      maxWidth: '100vw',
    },
    tableHead: {
      backgroundColor: '#3d474a',
    },
    innerAnalysisButton: {
      backgroundColor: '#3d474a',
    },
  });

interface AnalysisTableProps extends WithStyles<typeof styles> {
  tableData: AnalysisTableRow[];
  columns: Column[];
}

export default withStyles(styles)(AnalysisTable);
