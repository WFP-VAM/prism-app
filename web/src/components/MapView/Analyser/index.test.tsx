import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';

import Analyser from '.';
import { store } from '../../../context/store';

jest.mock('../Layers/LayerDropdown', () => 'mock-Layer-Dropdown');

test('renders as expected', () => {
  const rendered = render(
    <Provider store={store}>
      <Analyser />
    </Provider>,
  );
  return rendered
    .findByText('Run Analysis')
    .then(btn => btn.click()) // open analyser menu (default closed)
    .then(() => expect(rendered.container).toMatchSnapshot());
});
