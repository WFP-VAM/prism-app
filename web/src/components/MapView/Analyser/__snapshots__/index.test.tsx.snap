// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div>
  <div
    class="Analyser-analyser-1"
  >
    <mock-button
      color="primary"
      variant="contained"
    >
      <mock-barchart
        fontsize="small"
      />
      <mock-typography
        classname="Analyser-analyserLabel-2"
        variant="body2"
      >
        Run Analysis
      </mock-typography>
      <mock-arrowdropdown
        fontsize="small"
      />
    </mock-button>
    <div
      class="MuiBox-root MuiBox-root-13 Analyser-analyserMenu-3"
    >
      <div>
        <div
          class="Analyser-newAnalyserContainer-6"
        >
          <div
            class="Analyser-analyserOptions-5"
          >
            <mock-typography
              variant="body2"
            >
              Hazard Layer
            </mock-typography>
            <mock-layer-dropdown
              classname="Analyser-selector-10"
              placeholder="Choose hazard layer"
              title="Hazard Layer"
              type="wms"
            />
          </div>
          <div
            class="Analyser-analyserOptions-5"
          >
            <mock-typography
              variant="body2"
            >
              Statistic
            </mock-typography>
            <div
              class="MuiFormControl-root"
            >
              <div
                class="MuiFormGroup-root MuiFormGroup-row"
                role="radiogroup"
              >
                <label
                  class="MuiFormControlLabel-root"
                >
                  <span
                    aria-disabled="false"
                    class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-14 MuiRadio-root Analyser-radioOptions-7 PrivateSwitchBase-checked-15 Mui-checked"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <input
                        checked=""
                        class="PrivateSwitchBase-input-17"
                        name="statistics"
                        type="radio"
                        value="mean"
                      />
                      <div
                        class="PrivateRadioButtonIcon-root-18 PrivateRadioButtonIcon-checked-20"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                          />
                        </svg>
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root PrivateRadioButtonIcon-layer-19 MuiSvgIcon-fontSizeSmall"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                          />
                        </svg>
                      </div>
                    </span>
                    <span
                      class="MuiTouchRipple-root"
                    />
                  </span>
                  <mock-typography
                    classname="MuiFormControlLabel-label"
                    component="span"
                  >
                    Mean
                  </mock-typography>
                </label>
                <label
                  class="MuiFormControlLabel-root"
                >
                  <span
                    aria-disabled="false"
                    class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-14 MuiRadio-root Analyser-radioOptions-7"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <input
                        class="PrivateSwitchBase-input-17"
                        name="statistics"
                        type="radio"
                        value="median"
                      />
                      <div
                        class="PrivateRadioButtonIcon-root-18"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                          />
                        </svg>
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root PrivateRadioButtonIcon-layer-19 MuiSvgIcon-fontSizeSmall"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                          />
                        </svg>
                      </div>
                    </span>
                    <span
                      class="MuiTouchRipple-root"
                    />
                  </span>
                  <mock-typography
                    classname="MuiFormControlLabel-label"
                    component="span"
                  >
                    Median
                  </mock-typography>
                </label>
              </div>
            </div>
          </div>
          <div
            class="Analyser-analyserOptions-5"
          >
            <mock-typography
              variant="body2"
            >
              Baseline Layer
            </mock-typography>
            <mock-layer-dropdown
              classname="Analyser-selector-10"
              placeholder="Choose baseline layer"
              title="Baseline Layer"
              type="nso"
            />
          </div>
          <div
            class="Analyser-analyserOptions-5"
          >
            <mock-typography
              variant="body2"
            >
              Threshold
            </mock-typography>
            <mock-textfield
              classname="Analyser-numberField-11"
              error="false"
              id="filled-number"
              label="Min"
              type="number"
              value=""
              variant="filled"
            />
            <mock-textfield
              classname="Analyser-numberField-11"
              id="filled-number"
              label="Max"
              type="number"
              value=""
              variant="filled"
            />
          </div>
          <div
            class="Analyser-analyserOptions-5"
          >
            <mock-typography
              variant="body2"
            >
              Date
            </mock-typography>
            <div
              class="react-datepicker-wrapper"
            >
              <div
                class="react-datepicker__input-container"
              >
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline"
                >
                  <input
                    class="MuiInputBase-input MuiInput-input"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <mock-button
          classname="Analyser-innerAnalysisButton-8"
          disabled="true"
        >
          <mock-typography
            variant="body2"
          >
            Run Analysis
          </mock-typography>
        </mock-button>
      </div>
    </div>
  </div>
</div>
`;
